include/jemalloc/internal/arena_inlines_b.h

```diff
@@ -298,6 +298,61 @@ arena_dalloc_large(tsdn_t *tsdn, void *ptr, tcache_t *tcache, szind_t szind,
 	}
 }
 
+/*
+ * Internal.
+ *
+ * Find the region index of a pointer.
+*/
+JEMALLOC_ALWAYS_INLINE size_t
+arena_slab_regind_impl(div_info_t* div_info, szind_t binind,
+    edata_t *slab, const void *ptr) {
+	size_t diff, regind;
+
+	/* Freeing a pointer outside the slab can cause assertion failure. */
+	assert((uintptr_t)ptr >= (uintptr_t)edata_addr_get(slab));
+	assert((uintptr_t)ptr < (uintptr_t)edata_past_get(slab));
+	/* Freeing an interior pointer can cause assertion failure. */
+	assert(((uintptr_t)ptr - (uintptr_t)edata_addr_get(slab)) %
+	    (uintptr_t)bin_infos[binind].reg_size == 0);
+
+	diff = (size_t)((uintptr_t)ptr - (uintptr_t)edata_addr_get(slab));
+
+	/* Avoid doing division with a variable divisor. */
+	regind = div_compute(div_info, diff);
+	assert(regind < bin_infos[binind].nregs);
+	return regind;
+}
+
+/*
+* Checks whether ptr is already deallocated in the arena.
+* Fires safety check if it has been deallocated and returns true.
+*/
+JEMALLOC_ALWAYS_INLINE bool
+arena_tcache_dalloc_small_safety_check(tsdn_t *tsdn, void *ptr) {
+	if (!config_debug) {
+		return false;
+	}
+	edata_t *edata = emap_edata_lookup(tsdn, &arena_emap_global,
+	    ptr);
+	szind_t binind = edata_szind_get(edata);
+	div_info_t div_info = arena_binind_div_info[binind];
+	/* Calls the internal function arena_slab_regind_impl because the safety check
+	 * does not require a lock.
+	*/
+	size_t regind = arena_slab_regind_impl(&div_info, binind, edata, ptr);
```

Comment:
It looks to me that, we can simply call `arena_slab_regind` here? If so we don't even have to add the _impl anymore?

Reply:
`arena_slab_regind` requires a lock since its first variable taken is `arena_dalloc_bin_locked_info_t`. For the safety check, there is no need to acquire a lock, so we are using `arena_slab_regind_impl`, which does not require a lock.

Reply:
Ah I see. Now I remember our discussion. Thanks for the clarification.